### ISSUE TITLE
Prototype how funky groupings could look better

### ISSUE THREAD

**Issue by jhofman** [microsoft/datamations#105]
Here are some test specs and their videos:

## 100 points to 50/50

https://user-images.githubusercontent.com/15895337/138937144-ba792f6b-9353-4405-b449-7317afbb4864.mov

```
[
  {
    "height": 300,
    "width": 300,
    "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
    "meta": {
      "parse": "grid",
      "description": "Initial data"
    },
    "data": {
      "values": [
        {
          "n": 100
        }
      ]
    },
    "mark": {
      "type": "point",
      "filled": true
    },
    "encoding": {
      "x": {
        "field": "datamations_x",
        "type": "quantitative",
        "axis": null
      },
      "y": {
        "field": "datamations_y",
        "type": "quantitative",
        "axis": null
      }
    }
  },
  {
    "height": 300,
    "width": 300,
    "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
    "meta": {
      "parse": "grid",
      "description": "Group by group",
      "splitField": "group",
      "axes": false
    },
    "data": {
      "values": [
        {
          "group": "group1",
          "n": 50
        },
        {
          "group": "group2",
          "n": 50
        }
      ]
    },
    "mark": {
      "type": "point",
      "filled": true
    },
    "encoding": {
      "x": {
        "field": "datamations_x",
        "type": "quantitative",
        "axis": null
      },
      "y": {
        "field": "datamations_y",
        "type": "quantitative",
        "axis": null
      },
      "color": {
        "field": null,
        "type": "nominal"
      },
      "tooltip": [
        {
          "field": "group",
          "type": "nominal"
        }
      ]
    }
  }
] 
```

## 100 points to 48/52 (to see the crossover when the number isn't round to 10 / sqrt(1000)

https://user-images.githubusercontent.com/15895337/138937175-30a9f210-c5d1-4fe0-8921-52dd390ef08d.mov

```
[
  {
    "height": 300,
    "width": 300,
    "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
    "meta": {
      "parse": "grid",
      "description": "Initial data"
    },
    "data": {
      "values": [
        {
          "n": 100
        }
      ]
    },
    "mark": {
      "type": "point",
      "filled": true
    },
    "encoding": {
      "x": {
        "field": "datamations_x",
        "type": "quantitative",
        "axis": null
      },
      "y": {
        "field": "datamations_y",
        "type": "quantitative",
        "axis": null
      }
    }
  },
  {
    "height": 300,
    "width": 300,
    "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
    "meta": {
      "parse": "grid",
      "description": "Group by group",
      "splitField": "group",
      "axes": false
    },
    "data": {
      "values": [
        {
          "group": "group1",
          "n": 48
        },
        {
          "group": "group2",
          "n": 52
        }
      ]
    },
    "mark": {
      "type": "point",
      "filled": true
    },
    "encoding": {
      "x": {
        "field": "datamations_x",
        "type": "quantitative",
        "axis": null
      },
      "y": {
        "field": "datamations_y",
        "type": "quantitative",
        "axis": null
      },
      "color": {
        "field": null,
        "type": "nominal"
      },
      "tooltip": [
        {
          "field": "group",
          "type": "nominal"
        }
      ]
    }
  }
] 
```

## 100 points to 87/13 (again to see the crossover)

https://user-images.githubusercontent.com/15895337/138937434-a4bbc02b-04b3-49b3-94f7-f9bc059224fd.mov

```
[
  {
    "height": 300,
    "width": 300,
    "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
    "meta": {
      "parse": "grid",
      "description": "Initial data"
    },
    "data": {
      "values": [
        {
          "n": 100
        }
      ]
    },
    "mark": {
      "type": "point",
      "filled": true
    },
    "encoding": {
      "x": {
        "field": "datamations_x",
        "type": "quantitative",
        "axis": null
      },
      "y": {
        "field": "datamations_y",
        "type": "quantitative",
        "axis": null
      }
    }
  },
  {
    "height": 300,
    "width": 300,
    "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
    "meta": {
      "parse": "grid",
      "description": "Group by group",
      "splitField": "group",
      "axes": false
    },
    "data": {
      "values": [
        {
          "group": "group1",
          "n": 87
        },
        {
          "group": "group2",
          "n": 13
        }
      ]
    },
    "mark": {
      "type": "point",
      "filled": true
    },
    "encoding": {
      "x": {
        "field": "datamations_x",
        "type": "quantitative",
        "axis": null
      },
      "y": {
        "field": "datamations_y",
        "type": "quantitative",
        "axis": null
      },
      "color": {
        "field": null,
        "type": "nominal"
      },
      "tooltip": [
        {
          "field": "group",
          "type": "nominal"
        }
      ]
    }
  }
] 
```

## 100 points to 20 groups of 5

https://user-images.githubusercontent.com/15895337/138937478-7de5d5d0-f40c-4f75-a4d8-92a8aef5bb90.mov

```
[
  {
    "height": 300,
    "width": 300,
    "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
    "meta": {
      "parse": "grid",
      "description": "Initial data"
    },
    "data": {
      "values": [
        {
          "n": 100
        }
      ]
    },
    "mark": {
      "type": "point",
      "filled": true
    },
    "encoding": {
      "x": {
        "field": "datamations_x",
        "type": "quantitative",
        "axis": null
      },
      "y": {
        "field": "datamations_y",
        "type": "quantitative",
        "axis": null
      }
    }
  },
  {
    "height": 300,
    "width": 300,
    "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
    "meta": {
      "parse": "grid",
      "description": "Group by group",
      "splitField": "group",
      "axes": false
    },
    "data": {
      "values": [
        {
          "group": "group1",
          "n": 5
        },
        {
          "group": "group10",
          "n": 5
        },
        {
          "group": "group11",
          "n": 5
        },
        {
          "group": "group12",
          "n": 5
        },
        {
          "group": "group13",
          "n": 5
        },
        {
          "group": "group14",
          "n": 5
        },
        {
          "group": "group15",
          "n": 5
        },
        {
          "group": "group16",
          "n": 5
        },
        {
          "group": "group17",
          "n": 5
        },
        {
          "group": "group18",
          "n": 5
        },
        {
          "group": "group19",
          "n": 5
        },
        {
          "group": "group2",
          "n": 5
        },
        {
          "group": "group20",
          "n": 5
        },
        {
          "group": "group3",
          "n": 5
        },
        {
          "group": "group4",
          "n": 5
        },
        {
          "group": "group5",
          "n": 5
        },
        {
          "group": "group6",
          "n": 5
        },
        {
          "group": "group7",
          "n": 5
        },
        {
          "group": "group8",
          "n": 5
        },
        {
          "group": "group9",
          "n": 5
        }
      ]
    },
    "mark": {
      "type": "point",
      "filled": true
    },
    "encoding": {
      "x": {
        "field": "datamations_x",
        "type": "quantitative",
        "axis": null
      },
      "y": {
        "field": "datamations_y",
        "type": "quantitative",
        "axis": null
      },
      "color": {
        "field": null,
        "type": "nominal"
      },
      "tooltip": [
        {
          "field": "group",
          "type": "nominal"
        }
      ]
    }
  }
] 
```

_Originally posted by @sharlagelfand in https://github.com/microsoft/datamations/issues/102#issuecomment-952194920_

**Comment by jhofman** [comment]
The 20 groups of 5 look kind of strange at the moment. @giorgi-ghviniashvili, can you visually prototype how these might look more intuitive and then we'll figure out how to make that algorithmic in datamations code?

**Comment by giorgi-ghviniashvili** [comment]
@jhofman @sharlagelfand I thought that maybe we can first collapse 10 groups into 5 and then animate right? Like this:

https://user-images.githubusercontent.com/6615532/140492806-07bdf9af-d1c1-4643-8a0e-f99556ded3bb.mov

But as I searched, we can utilize gemini2 to further instruct intermediate animations.. Check this out: https://uwdata.github.io/gemini2-editor/

**Comment by jhofman** [comment]
We all liked the idea of squashed vertical for these groups (whitespace on top and bottom) for these many small groups. 

Maybe the heuristic here is that you don't want too much whitespace between points in a group in some "absolute" sense (relative to the size of the bounding box)?

@giorgi-ghviniashvili, can you play around with this and also with the "faceted grids" arranged as we drew on paper?

**Comment by giorgi-ghviniashvili** [comment]
Adjusted y axis domain so it has good amount of padding. 

https://user-images.githubusercontent.com/6615532/140720318-37d6cf43-be11-44ff-a318-83a90eba4033.mov

**Comment by jhofman** [comment]
Small note here that if we go to a group_by with a skewed number of points in one group, it can get hard to see the points in that large group.

Example here:

https://github.com/microsoft/datamations/issues/114#issuecomment-971887510

**Comment by jhofman** [comment]
@giorgi-ghviniashvili, can you think a bit about heuristics for min and max distance between points in a group, and how that could affect row/column layouts?

**Comment by giorgi-ghviniashvili** [comment]
@jhofman here is how I calculate rows now:

* first find square root of max group
* then if it is greater than the `distance`, then compute rows based on distance. 

```javascript
    // calculate rows: considering spec height for rows calculation

    const n = d3.max(vegaLiteSpecs[0].data.values, d => d.n);
    let rows = Math.ceil(Math.sqrt(n));

    const gap = 2;
    const distance = 6 + gap;

    const {height: firstSpecHeight} = vegaLiteSpecs[0].spec || vegaLiteSpecs[0];

    if (firstSpecHeight / rows > distance) {
      rows = Math.floor(firstSpecHeight / distance)
    }

    // end of rows calculation
```

**Comment by jhofman** [comment]
@giorgi-ghviniashvili: can you re-render the test cases at the top of this issue using this new heuristic and see how it looks?

**Comment by giorgi-ghviniashvili** [comment]
Here is how 20 groups look like. Better than top example:

https://user-images.githubusercontent.com/6615532/143875212-83fefdce-d129-4658-82b2-28665e11166c.mov

We can also apply max number of rows if you want, to avoid 3 columns. But for me this is ok, without max number of rows.

**Comment by giorgi-ghviniashvili** [comment]
52/48 is odd for me:

https://user-images.githubusercontent.com/6615532/143875696-08010128-8571-4c5b-975e-fa0ad2b60250.mov

**Comment by giorgi-ghviniashvili** [comment]
50/50 is same as 48. Odd. Maybe we can apply some threshold and if number of circles are greater than this threshold, say 200 or 300, then calculate row number based on min distance in order to avoid overlapping, otherwise, just use square root way. What do you think? @jhofman . If there are few points in total: 100 or less, it makes the viz very narrow and vertical. Don't like it.

**Comment by jhofman** [comment]
agreed that the newer version gets a little weird and might not generalize well.

seems like we may want to go with something more like [this](https://github.com/microsoft/datamations/issues/105#issuecomment-962980965) but try to adjust the gap between points in subsequent frames.

@giorgi-ghviniashvili will look at this for the next mtg.

**Comment by jhofman** [comment]
this will be top priority for @giorgi-ghviniashvili.

once done, @sharlagelfand will merge in ALL THE THINGS.

**Comment by giorgi-ghviniashvili** [comment]
So the original issue why we started to consider `gap` for `rows` calculating was the overlapping circles: 

![image](https://user-images.githubusercontent.com/6615532/144507385-1fb1d31c-e31f-4144-9533-e7a1b5d6f0b4.png)

We could increase rows and have less columns to avoid overlapping. 

For this have in mind, here is my solution:

* if width (within a group / facet) divided by columns is very small that causes overlap, let's increase rows, to take up all the space vertically and reduce columns. 

Here is the code and result:

```javascript
let maxCols = Math.ceil(d3.max(specValues, d => d.n) / rows);

// if width divided by maxCols is less than 5, 
// then take up all vertical space to increase rows and reduce columns 
if (specWidth / maxCols < 5) {
  rows = Math.floor(specHeight / distance);
  maxCols = Math.ceil(d3.max(specValues, d => d.n) / rows);
}
```

Result:
![image](https://user-images.githubusercontent.com/6615532/144505752-b32fb005-6877-4722-8faf-2ae99b6913b0.png)

* But for this case:

![image](https://user-images.githubusercontent.com/6615532/144505881-662960b3-2466-4d87-9162-a519c7663325.png)

It will stay same squared root rows, because we don't really need to adjust the rows here. They don't overlap!!! 

![image](https://user-images.githubusercontent.com/6615532/144507173-f325cf49-230e-4640-b000-0c3851c47d90.png)

![image](https://user-images.githubusercontent.com/6615532/144507224-056a04f0-8ef6-4c20-ac78-0fb5bdf817ea.png)

So to sum up, the funky group is still there. I guess increase of row number can not fix that, but at least we can MERGE.

**Comment by sharlagelfand** [comment]
Closed by #118
